Hash routing to named element

surayya-MS opened the pull request:
# Hash routing to named element


## Description

Implemented navigation to the named element via anchor element or Navlink component with href containing '#' or using NavigationManager to navigate to url containing '#'.
```razor
<a href="#test1">Anchor to test1</a>
<NavLink href="#test1">NavLink to test1</NavLink>
<a href="/counter#test2">Anchor to test1</a>
```
There are 2 scenarios:

1. Scrolling to the element on the same page
   * Anchor element (or NavLink component which renders as anchor) with href  starting with '#' 
   * Anchor with href with the same url but with changed hash
   
   For this scenario I changed `NavigationManager.ts` `performInternalNavigation` which is called in the method that overrides default behavior of anchor elements to handle this scenario too.

3. Internal navigation
   * Anchor element with href that contains '#' linking to another page
   *  using NavigationManager to navigate to another page or even the same page
   
   For this scenario I changed `Router` `OnAfterRenderAsync` to check the url and if it contains '#' then call a js script to scroll to the element.

Also I had to change behavior of `FocusOnNavigate` component because the page would flicker - scroll to the named element and then scroll to the `FocusOnNavigate` Selector. The ts script that `FocusOnNavigate` uses now prevents scrolling when focusing on the selector.


Fixes #8393

surayya-MS committed:
implemented scrolling to location hash

surayya-MS committed:
BlazorServerApp test hash

surayya-MS committed:
Merge branch 'main' into hashNav

surayya-MS committed:
Revert "BlazorServerApp test hash"

This reverts commit 65133e09d7609d943da9bb9142ee10d14e29e049.

surayya-MS committed:
remove spaces

surayya-MS committed:
change Router OnAfterRenderAsync

surayya-MS committed:
fix

surayya-MS committed:
small fix

javiercn commented on src/Components/Components/src/Microsoft.AspNetCore.Components.csproj:
```diff
@@ -22,6 +22,7 @@
   <ItemGroup>
     <Reference Include="Microsoft.Extensions.Logging.Abstractions" />
     <Reference Include="Microsoft.Extensions.DependencyInjection.Abstractions" />
+    <Reference Include="Microsoft.JSInterop" />
```
Is this because of the router?

surayya-MS replied:
yes

javiercn replied:
We want to avoid having a reference to Microsoft.JSInterop here. I would suggest you check how enablenavigationinterception works to see if we can follow a similar approach.

@SteveSandersonMS do you have other thoughts?

SteveSandersonMS replied:
The way @surayya-MS has addressed this using `IScrollToLocationHash` looks good to me (if we are implementing it on the .NET side, but see comments below).

javiercn reviewed (commented):
The change looks reasonable, but I think we need E2E tests for it, and we need to address the issue with the layering, as we don't want to add a reference to JS interop directly to Microsoft.AspNetCore.Components

surayya-MS committed:
implemented IScrollToLocationHash; added e2e tests

SteveSandersonMS commented on src/Components/Components/src/Routing/Router.cs:
```diff
@@ -277,14 +279,19 @@ private void OnLocationChanged(object sender, LocationChangedEventArgs args)
         }
 
-        return Task.CompletedTask;
+        await ScrollToLocationHash.ScrollToLocationHash(_locationAbsolute);
```
Does this mean that every time the `<Router>` renders, it will re-scroll to the current URL hash? If so that seems like incorrect behavior because it could result in the user's scroll position changing unexpectedly, unrelated to navigation.

To clarify, as far as I'm aware, the router could re-render at any time - not necessarily just because some navigation happened. It would happen any time time component containing the `<Router>` re-renders, which can be for arbitrary application-specific reasons.

If this has been accounted for somehow then sorry for the confusion, but please clarify if so!

surayya-MS replied:
> Does this mean that every time the <Router> renders, it will re-scroll to the current URL hash?

Yes, it does.

Thank you! I didn't know that `Router` could re-render at any time.

I just found a better way of achieving the same without changing `Router.OnAfterAsync`. The challenge was to catch a moment when the page is rendered. [Here](https://github.com/dotnet/aspnetcore/blob/main/src/Components/Web.JS/src/Rendering/Renderer.ts#L90) I will add scrolling to the element.

SteveSandersonMS commented on src/Components/Web.JS/src/Services/NavigationManager.ts:
```diff
@@ -76,8 +85,18 @@ export function attachToEventDelegator(eventDelegator: EventDelegator): void {
-      const absoluteHref = toAbsoluteUri(href);
+      const anchorHref = anchorTarget.getAttribute('href')!;
+
+      if (anchorHref.startsWith('#') && anchorHref.length > 1) {
```
This looks like it would intercept `href="#something"` but not for `href="current/page#something"`.

We should clarify what the intended behavior is in the case of `href="current/page#something"`. Two possibilities:

1. We should treat this as a navigation to `current/page` even though that's already where you are, and *then* scroll to `#something`. This involves going through the .NET-side routing process and possibly notifying the page that it has new parameters (which may be the same as its previous ones), possibly causing it to re-render.
2. Or, we should not treat this as a navigation at all, since we are already at `current/page`, and instead just short-circuit and scroll to `#something`. The .NET side would not receive any notification and hence wouldn't re-render anything.

My guess from seeing the code here is that the implementation in this PR does option (1). Is that the desired design? To be totally honest I think most developers would not expect there to be a difference between navigating to `#something` and `current/page#something` if you are already at `current/page`.

So this is an open design choice and we can choose whatever we want, but I'm guessing option (2) may be more preferable for most developers.

SteveSandersonMS replied:
I just checked and with regular native `<a>` tags, the behavior of navigating to `current/page#something` is identical to navigating to `#something` (as in, it does not cause a new page load), so we probably should treat the two as identical in the same way.

SteveSandersonMS replied:
We might need to have some slightly more subtle rule, like:

 * If you navigate to the existing URL where only the hash has been added/changed **and** that anchor is already in the DOM, then we just scroll to it and don't notify the .NET side. That's the closest to native HTML behavior.
 * But, if the anchor isn't already in the DOM, we don't short-circuit the navigation. We treat it as any other navigation and hence notify the .NET side and begin the whole process of waiting for the anchor to be added to the DOM asynchronously as per [this](https://github.com/dotnet/aspnetcore/pull/47320#issuecomment-1486593691).

Otherwise, if we just short-circuit all navigations to `current/page#hash` even when the anchor isn't in the page, that click will never have any effect.

SteveSandersonMS commented on src/Components/Web.JS/src/Services/NavigationManager.ts:
```diff
@@ -53,6 +54,14 @@ function setHasLocationChangingListeners(hasListeners: boolean) {
   hasLocationChangingEventListeners = hasListeners;
 }
 
+function scrollToElement(id : string) : void {
```
This only accounts for `<elem id="...">` but would also need to account for `<a name="...">` (https://www.tagindex.net/html/link/a_name.html).

The parameter should be an anchor name, not an element ID. I'm not sure whether it should prioritize elements by ID over anchors by name - would it be possible to check how native `<a>` tags work and follow the same prioritization?

SteveSandersonMS commented:
I think there's also a third scenario we should consider. Or at least, a variation of scenario 2 in the PR description:

 2. (B) Internal navigation with async data loading
      * Anchor tag goes to `chat#message1234`
      * The `chat` page renders a series of tags like `<a name="message1234">...</a>` **but they are rendered asychronously** because of some loading process that occurs first

The question is, how long is the async delay allowed to be? How do we know when to stop waiting? In the existing implementation in this PR, it has to be zero delay - we only support navigating to anchors that exist synchronously after the navigation. But I think in a very large number of cases, people have no choice but to load stuff asynchronously. At the same time, we don't want to wait indefinitely for the anchor tag to be added, since maybe the user has already started doing unrelated scrolling and will be inconvenienced by the scroll position changing unexpectedly later.

Scenarios like this make me thing that a purely client-side implementation might be simpler and more flexible. That is:

 * When the JS-side code intercepts a navigation, if it contains a `#hash`, it stores that and sets some kind of timer or other cancellable token (and on all navigations, cancel any prior pending state)
 * After each renderbatch is applied, the JS-side code checks if there is a pending `#hash` and if so, tries to find it in the page.
   * If it's found, scroll to that point and clear all this temporary `#hash` state away
   * If not, just keep waiting
 * If the timer expires, clear the temporary `#hash` state
 * Also, during the initial page startup process, if the initial URL contains a `#hash`, set up the equivalent temporary state

Then none of the .NET-side changes would be necessary. This is at least more general than the existing solution, at the cost that while a `#hash` is pending we have to search the DOM for it after every renderbatch (which I think is OK because this doesn't happen often and in most cases it would stop after the next renderbatch).

The one issue is extreme cases where it takes, say, 20+ seconds for the async loading process to complete. We probably don't want to wait that long. So we'd have to accept the limitation that scrolling to the hash is not guaranteed and will only happen if you render the corresponding content within, say, 5 seconds. This does get us into the realm of making subjective judgements about UX but I think that's unavoidable.

What do you think? Was this sort of approach considered already?

surayya-MS committed:
deleted IScrollToLocationHash; implemented scrolling on the client side

surayya-MS commented:
I will add more e2e tests soon

surayya-MS committed:
small fix

surayya-MS committed:
small fix

surayya-MS committed:
fix focusOnNavigate

surayya-MS committed:
change timer to 5 seconds

SteveSandersonMS commented on src/Components/Web/src/DomWrapperInterop.cs:
```diff
@@ -9,5 +9,5 @@ internal static class DomWrapperInterop
     public const string Focus = Prefix + "focus";
 
-    public const string FocusBySelector = Prefix + "focusBySelector";
+    public const string FocusOnNavigate = Prefix + "focusOnNavigate";
```
I'm not sure I understand this change. The existing `FocusOnNavigate` component will still need to call `focusBySelector`, won't it?

surayya-MS replied:
yes, `focusOnNavigate` will call `focusBySelector`

SteveSandersonMS commented on src/Components/Web.JS/src/DomWrapper.ts:
```diff
@@ -22,6 +23,12 @@ function focus(element: HTMLOrSVGElement, preventScroll: boolean): void {
+function focusOnNavigate(selector: string): void {
+  if ( !(location.hash.length > 1 && elementExists(location.hash.slice(1))) ) {
+    focusBySelector(selector);
+  }
```
It looks like this logic suppresses `FocusOnNavigate` if the URL contains a hash. I haven't actually tried this out so am not sure what the exact actual behavior is but in general would expect `focus` to be independent to `scrollIntoView`.

Does it work to leave `FocusOnNavigate` to do its existing behavior regardless of the URL hash?

The reason is that `FocusOnNavigate` is an accessibility feature that causes a page title to be announced. I would expect people still want that to happen even if we're also auto-scrolling to some unrelated element based on the URL hash.

surayya-MS replied:
> It looks like this logic suppresses FocusOnNavigate if the URL contains a hash.
Yes it suppresses if the url contains hash and the element is in the page. 

> Does it work to leave FocusOnNavigate to do its existing behavior regardless of the URL hash?

No, the focus script is called after the scroll script so when navigating to another url with hash doesn't work.

SteveSandersonMS replied:
I don't get why `FocusOnNavigate` needs to be affected by this. Why do the following two things interact?

1. The `<FocusOnNavigate>` component performs a JS interop call after certain renders that results in a call to `someDomElement.focus()`
2. The new logic you've added in `Renderer.ts` runs after each renderbatch and may result in a call to `someOtherDomElement.scrollIntoView()`

What causes these to interfere with each other? Focus and scroll are two different things so I would expect them to work independently.

Sorry if I'm missing something obvious!

surayya-MS replied:
`focus()` also affects the scroll position. https://developer.mozilla.org/en-US/docs/Web/API/HTMLElement/focus
I think I'd better use `preventScroll` option rather than not focusing at all. Thanks!

SteveSandersonMS commented on src/Components/Web.JS/src/DomWrapper.ts:
```diff
@@ -35,3 +42,23 @@ function focusBySelector(selector: string): void {
   }
 }
+
+function elementExists(identifier : string) : boolean {
```
Sorry to be nitpicky, but is it possible to tweak the code style to match the other TypeScript code? For example, this would be:

```suggestion
function elementExists(identifier: string): boolean {
```

I know it's only a minor detail 😄

surayya-MS replied:
Sure. I will fix in other places too.

SteveSandersonMS commented on src/Components/Web.JS/src/DomWrapper.ts:
```diff
@@ -35,3 +42,23 @@ function focusBySelector(selector: string): void {
 }
+
+function elementExists(identifier : string) : boolean {
+  let element : HTMLElement | null = null;
+  
+  element = document.getElementById(identifier);
+  
+  if (!element) {
+
+    let elements = document.getElementsByName(identifier);
+    if (elements.length > 0) {
+      element = elements[0];     
+    }
+  }
+  
+  if (element) {
+    return true;
+  }
+
+  return false;
```
TypeScript/JavaScript is a pretty forgiving language when it comes to null handling, so this could all be reduced to:

```suggestion
  const element = document.getElementById(identifier)
      || document.getElementsByName(identifier)[0];
  return !!element;
```

Sidenote: we generally prefer `const` over `let` whenever possible, as this indicates that the variable won't be overwritten later. It's quite different to the meaning of `const` in C#.

SteveSandersonMS commented on src/Components/Web.JS/src/Rendering/Renderer.ts:
```diff
@@ -13,6 +13,8 @@ interface BrowserRendererRegistry {
 const browserRenderers: BrowserRendererRegistry = {};
 let shouldResetScrollAfterNextBatch = false;
+let elementToScrollTo : string | null = null;
+let scrollToElementTimeout : NodeJS.Timeout;
```
Could we avoid referencing `NodeJS` types here? The reason is this code is not going to run in Node, so if we use Node types it's quite difficult to reason about whether or not it will work correctly in a browser.

I know in this case that it would work because browser also support `setTimeout` but in browsers the result type for that is just `number` rather than `NodeJS.Timeout`.

SteveSandersonMS commented on src/Components/Web.JS/src/Rendering/Renderer.ts:
```diff
@@ -88,7 +90,46 @@ export function renderBatch(browserRendererId: number, batch: RenderBatch): void
+    clearTimeout(scrollToElementTimeout); 
+
+    // We found the element on the page and we don't want to scroll to the top of the page on the next batch
+    shouldResetScrollAfterNextBatch = false; 
```
This looks good!

SteveSandersonMS commented on src/Components/Web.JS/src/Rendering/Renderer.ts:
```diff
@@ -88,7 +90,46 @@ export function renderBatch(browserRendererId: number, batch: RenderBatch): void
+}
+
+export function scrollToElement(identifier : string) : boolean {
+  let element : HTMLElement | null = null;
+
+  element = document.getElementById(identifier);
+  
+  if (!element) {
+    let elements = document.getElementsByName(identifier);
+    if (elements.length > 0) {
+      element = elements[0];     
+    }
+  }
```
```suggestion
  const element = document.getElementById(identifier)
    || document.getElementsByName(identifier)[0];
```

SteveSandersonMS commented on src/Components/Web.JS/src/Services/NavigationManager.ts:
```diff
@@ -76,8 +76,18 @@ export function attachToEventDelegator(eventDelegator: EventDelegator): void {
-      const absoluteHref = toAbsoluteUri(href);
+      let anchorHref = anchorTarget.getAttribute('href')!; 
+      if (anchorHref.startsWith('#')) {
+        anchorHref = location.hash.length > 1 ? location.href.replace(location.hash, anchorHref) : location.href + anchorHref;
```
Rather than using `location.href.replace` and having two different cases for whether or not the URL already has a hash, could this be made more easy to follow with code like the following?

```
anchorHref = `${location.origin}${location.pathname}${location.search}${location.hash}`;
```

That way, the browser is dealing with most of the parsing details for you and you don't have to distinguish different cases.

SteveSandersonMS commented on src/Components/Web.JS/src/Services/NavigationManager.ts:
```diff
@@ -87,6 +97,24 @@ export function attachToEventDelegator(eventDelegator: EventDelegator): void {
+  const hashIndex = absoluteHref.indexOf('#');
+  var hash = absoluteHref.substring(hashIndex); 
+
+  const urlInBrowser = location.hash.length > 1 ? location.href.replace(location.hash, hash) : location.href + hash;
```
Could you clarify why we build `urlInBrowser` here instead of just using `absoluteHref` directly? Isn't that already the exact URL we want to use for `history.pushState`?

surayya-MS replied:
You are right! I completely missed to change this piece of code 😄  it previously was somewhere else where there was no absolutehref. Thanks!

SteveSandersonMS commented on src/Components/Web.JS/src/Services/NavigationManager.ts:
```diff
@@ -87,6 +97,24 @@ export function attachToEventDelegator(eventDelegator: EventDelegator): void {
 
+function isSamePageWithHash(absoluteHref : string) : boolean {
+  const hashIndex = absoluteHref.indexOf('#');
+  // Excluding case hash="#"
```
Could you extend the comment to say why we want to exclude this case? It's not obvious to me at least :)

SteveSandersonMS commented on src/Components/Web.JS/src/Services/NavigationManager.ts:
```diff
@@ -152,6 +180,13 @@ async function performInternalNavigation(absoluteInternalHref: string, intercept
+  if ( hashIndex > -1 && absoluteInternalHref.length > hashIndex + 1) {
+    var identifier = absoluteInternalHref.substring(hashIndex+1);
+    setTimeoutToScrollToElement(identifier);
+  }
```
Minor code style issues: leading space after `(` on L185, use `const` instead of `var`

SteveSandersonMS replied:
Also inconsistent spacing around `+`

SteveSandersonMS commented:
I'm not sure if we've covered the following and suspect we might have missed it. The new logic deals with navigations triggered by people clicking on links, but does it also deal with navigations triggered in these ways?

1. `NavigationManager.NavigateTo` called from C# code
2. `Blazor.navigateTo` called from JS code

It's quite possible that 2 is just another way to trigger 1, so it might be that only 1 needs to be handled.

This will be necessary to support people who want to trigger scrolling to a named anchor as part of their `OnAfterRenderAsync` logic. Would it be possible to add an E2E test case showing a component that triggers scrolling to a named anchor from its `OnAfterRenderAsync` method by calling `navigationManager.NavigateTo`?

surayya-MS commented:
> I'm not sure if we've covered the following and suspect we might have missed it. The new logic deals with navigations triggered by people clicking on links, but does it also deal with navigations triggered in these ways?
> 
> 1. `NavigationManager.NavigateTo` called from C# code
> 2. `Blazor.navigateTo` called from JS code
> 
> It's quite possible that 2 is just another way to trigger 1, so it might be that only 1 needs to be handled.
> 
> This will be necessary to support people who want to trigger scrolling to a named anchor as part of their `OnAfterRenderAsync` logic. Would it be possible to add an E2E test case showing a component that triggers scrolling to a named anchor from its `OnAfterRenderAsync` method by calling `navigationManager.NavigateTo`?

Yes, this is also covered. I will add this case to e2e tests

surayya-MS commented:
I just checked again the case when calling `NavigationManager.NavigaTo()` same url + hash. it triggers rendering. I will have to change the script a little bit.

surayya-MS committed:
Update src/Components/Web.JS/src/DomWrapper.ts

Co-authored-by: Steve Sanderson <SteveSandersonMS@users.noreply.github.com>

surayya-MS committed:
Update src/Components/Web.JS/src/DomWrapper.ts

Co-authored-by: Steve Sanderson <SteveSandersonMS@users.noreply.github.com>

surayya-MS committed:
Update src/Components/Web.JS/src/Rendering/Renderer.ts

Co-authored-by: Steve Sanderson <SteveSandersonMS@users.noreply.github.com>

surayya-MS committed:
Revert "Update src/Components/Web.JS/src/Rendering/Renderer.ts"

This reverts commit 58b7a5512132b80b21410726f09ee94c83925626.

surayya-MS committed:
Revert "Update src/Components/Web.JS/src/DomWrapper.ts"

This reverts commit 4ccbce47a0dc506e4c978dacf72a5ac98f988741.

surayya-MS committed:
Revert "Update src/Components/Web.JS/src/DomWrapper.ts"

This reverts commit a52fc5bd70811baabea591f89b069309b787eb8f.

surayya-MS committed:
Revert "change timer to 5 seconds"

This reverts commit 0fdf00977809cfd5eea096e30a5cac83be0d8f7b.

surayya-MS committed:
Revert "fix focusOnNavigate"

This reverts commit ea8c152446970615be33e3b52e95ab521d2218ea.

surayya-MS committed:
Revert "small fix"

This reverts commit dab0f04a078129aa0d0489510af408f6a37b7d89.

surayya-MS committed:
Revert "small fix"

This reverts commit 82e3726e3bf34bfc26524bb015ea312d31dd7755.

surayya-MS committed:
Revert "deleted IScrollToLocationHash; implemented scrolling on the client side"

This reverts commit 8763e5b3b738fb8cc60fd0fb0501d9ea630ccaa4.

surayya-MS committed:
back to Router design;
suppress scrolling when Router re-renders;
add more e2e tests

surayya-MS committed:
Merge branch 'main' into hashNav

SteveSandersonMS committed:
Add pathbase support to Photino WebView so we can run the router E2E tests there

SteveSandersonMS committed:
implemented WebViewScrollToLocationHash

SteveSandersonMS committed:
Add pathbase support to Photino WebView so we can run the router E2E tests there

MackinnonBuck commented on src/Components/Web.JS/src/Services/NavigationManager.ts:
```diff
@@ -87,6 +105,23 @@ export function attachToEventDelegator(eventDelegator: EventDelegator): void {
+  return hashIndex > -1 && location.href.replace(location.hash, '') === absoluteHref.substring(0, hashIndex);
+}
+
+function performScrollToElementOnTheSamePage(absoluteHref : string): void {
+  history.pushState({}, "", absoluteHref);
```
In the [one other place](https://github.com/dotnet/aspnetcore/blob/90efca2f6a7682e0c952f26dfe1147c7766cd9d2/src/Components/Web.JS/src/Services/NavigationManager.ts#L199) we already call `pushState`, we include an `_index` property in the object that we pass as the `state` parameter. Could we do that here as well? Feel free to make a helper function to reduce code duplication.

---

Just for some context (and this really isn't that relevant to your PR so feel free to ignore, but I'm writing it for completeness), the reason we store a history "index" is to support an edge case in the "navigation cancellation" feature. When enabled, that feature has the requirement that we need to be able to "revert" navigations as soon as they happen, then get asynchronous confirmation from the .NET side to either replay or abort the navigation.

The edge case is when the user's navigation skips over multiple history entries (e.g., via the right-click menu for the browser's forward/back buttons). We need a way of detecting how far back/forward the navigation went so we know how far to move through the history stack to revert the navigation. An "index" in each history entry allows us to do the math required to perform that operation.

Without persisting the index here, I _think_ a bug would occur when the "location changing" event is enabled and the user navigates back to the history entry caused by the scroll navigation. An undefined `_index` is interpreted as an index of `0`, which could send the user way further back in history than they intended.

surayya-MS replied:
Thanks for ringing that up!

surayya-MS committed:
Merge branch 'hashNav' of https://github.com/surayya-MS/aspnetcore into hashNav

surayya-MS committed:
Merge branch 'hashNav' of https://github.com/surayya-MS/aspnetcore into hashNav

surayya-MS committed:
small fix

surayya-MS committed:
include _index in history.push()

SteveSandersonMS commented on src/Components/Components/test/Routing/RouterTest.cs:
```diff
@@ -220,6 +221,16 @@ public Task EnableNavigationInterceptionAsync()
 
+    internal sealed class TestScrollToLocationHash : IScrollToLocationHash
+    {
+        public static readonly TestScrollToLocationHash Instance = new TestScrollToLocationHash();
```
Looks like this field isn't used and could be removed

SteveSandersonMS commented on src/Components/Server/src/Circuits/RemoteScrollToLocationHash.cs:
```diff
@@ -0,0 +1,47 @@
+                "attempted during prerendering or while the client is disconnected.");
+        }
+
+        var hashIndex = locationAbsolute.IndexOf("#", StringComparison.CurrentCultureIgnoreCase);
```
```suggestion
        var hashIndex = locationAbsolute.IndexOf("#", StringComparison.Ordinal);
```

The `#` symbol in a URL isn't culture-specific and can't vary by case, so it would be marginally more efficient to search for it without accounting for culture and case-insensitivity.

surayya-MS replied:
Thanks! I applied same thing in `WebAssemplyScrollToLocationHash` and `WebViewScrollToLocationHash`

SteveSandersonMS commented on src/Components/Web.JS/src/DomWrapper.ts:
```diff
@@ -5,7 +5,7 @@ import '@microsoft/dotnet-js-interop';
 export const domFunctions = {
   focus,
-  focusBySelector,
+  focusOnNavigate
```
I don't think this change is required any more - the old name was slightly more descriptive since in theory this could be called at any time, not just on navigation. However I don't really mind if you prefer the new name!

SteveSandersonMS commented on src/Components/Web.JS/src/DomWrapper.ts:
```diff
@@ -22,7 +22,12 @@ function focus(element: HTMLOrSVGElement, preventScroll: boolean): void {
-function focusBySelector(selector: string): void {
+function focusOnNavigate(selector: string): void {
+  const preventScroll = location.hash.length > 1 && elementExists(location.hash.slice(1));
+  focusBySelector(selector, preventScroll);
```
I would suggest we make this *always* use `preventScroll`, regardless of whether there's a hash in the URL. The `<FocusOnNavigate>` component was never intended to affect scroll position, and historically wouldn't do so based on how the project templates were set up.

I think it's better to just `preventScroll` in all cases here because otherwise the behavior becomes less predictable - there's a conflict between what the hash routing is trying to do, and what `<FocusOnNavigate>` would be trying to do, and it would be hard to know for sure which one the developer wants to win this fight. It will be clearer if `<FocusOnNavigate>` *only* causes focusing and never causes scrolling, then there is no ambiguity.

surayya-MS replied:
I agree. I changed `focusBySelector` to always preventscroll and removed `focusOnNavigate` since there is no longer need for it. I also removed 2 test in e2e that were schecking scrolling when FocusOnNavigate selector is on the page.

SteveSandersonMS reviewed (commented):
This is really great! Thanks so much for continuing to sort out all the details and being flexible about the overall approach.

I would have approved this now (since this is 99% done) but just noticed that auto-merge is enabled. There are just a few minor details to resolve as per the other comments, and so I don't want to cause this to merge too soon 😄

SteveSandersonMS commented:
Also, superb work on the E2E tests. Those look really good.

surayya-MS committed:
Update src/Components/Server/src/Circuits/RemoteScrollToLocationHash.cs

Co-authored-by: Steve Sanderson <SteveSandersonMS@users.noreply.github.com>

surayya-MS commented on src/Components/Server/src/Circuits/RemoteScrollToLocationHash.cs:
```diff
@@ -0,0 +1,47 @@
+        {
+            var elementId = locationAbsolute[(hashIndex + 1)..];
+
+            await _jsRuntime.InvokeVoidAsync(Interop.ScrollToElement, elementId).AsTask();
```
Yes, I don't need that. I forgot to remove it after I made this method async. Thanks!

surayya-MS committed:
1. Use StringComparison.Ordinal when searching for '#'
2. Prevent scrolling in focusOnNavigate
3. Removed test that checks scrolling to a named element with FocusOnNavigate

surayya-MS committed:
Merge branch 'hashNav' of https://github.com/surayya-MS/aspnetcore into hashNav

SteveSandersonMS reviewed (approved):
Great!
